subxt/src/backend/unstable/mod.rs

```diff
@@ -145,34 +148,51 @@ impl<T: Config> UnstableBackend<T> {
         I: IntoIterator<Item = follow_stream_unpin::BlockRef<T::Hash>> + Send + 'static,
         <I as IntoIterator>::IntoIter: Send,
     {
-        let sub_id = get_subscription_id(&self.follow_handle).await?;
-        let sub_id = Arc::new(sub_id);
         let methods = self.methods.clone();
-        let headers = self.follow_handle.subscribe().events().flat_map(move |ev| {
-            let sub_id = sub_id.clone();
-            let methods = methods.clone();
-
-            let block_refs = f(ev).into_iter();
+        let follow_handle = self.follow_handle.clone();
+        let sub_id = get_subscription_id(&follow_handle).await?;
 
-            futures::stream::iter(block_refs).filter_map(move |block_ref| {
-                let sub_id = sub_id.clone();
+        // TODO(niklasad1): get rid of the Arc<RwLock> if possible...
+        let headers = self
+            .follow_handle
+            .subscribe()
+            .scan(Arc::new(RwLock::new(sub_id)), |id, v| {
+                let sub_id = id.clone();
+                async { Some((sub_id, v)) }
+            })
+            .filter_map(|(sub_id, v)| async {
+                match v {
+                    FollowStreamMsg::Event(ev) => Some((sub_id, ev)),
+                    FollowStreamMsg::Ready(id) => {
+                        *sub_id.write().unwrap() = id;
+                        None
+                    }
+                }
+            })
+            .flat_map(move |(sub_id, ev)| {
```

Comment:
This `stream_headers` is used to stream new/best/finalized blocks. One thing we want to be sure of is that no important blocks are missed (much like in the legacy backend) orrr we give an error to the user if they might be, so that they know it's possible.

It's a bit easier to pick up where we left off in the unstable backend; we can keep track of the latest finalized block we send to the user, and then on reconnect, if that block is in the `Initialized` event then we pick up from there and stream all the blocks after it, else we return the error to the user so that it's up to them if they want to resume (and miss blocks) or stop.

Reply:
IIUC, we could probably guard around the following edge-cases:

```
// Finalized blocks T0:
// We always keep track of the last one, in this case D.
chain: A -> B -> C -> D

// We get disconnected at T1 and reconnected at T1 + K
// Reconnect to chainHead will give us a new subID and the Initialized event
FollowEvent::Initialized { finalized_blocks: [C, D, E, F]}

// We search for our last finalized block D in the event Initialized
// We find it and continue normally. However if it is not present, we return FollowEvent::Stop
// because we cannot uphold the chainHead guarantee

```

Reply:
> Yes that makes sense but what I'm trying to explain is that the follow_stream stream impl doesn't keep state it just passes the events to next layer and it's currently no way to pass that DisconnectWillReconnect to the next layer except returning an error on the stream which in turn will emit an error on the backend driver.

I think it will pass a `stop` event though on disconnect, and then a new `initialized` event etc on reconnect. So as Alex said, you can do something like:

- Keep track of the last finalized block
- Wait for new `initialized` event.
    - If the last finalized block is in the list of finalized blocks in the `initialized` event then we can emit the remaining finalized blocks from that event and continue as normal.
    - If the last finalized block is **not** in the `initialized` list, then we emit some error and end the stream, because otherwise the user would miss blocks and not know about it.